h/groups/search.py

```diff
@@ -0,0 +1,23 @@
+# -*- coding: utf-8 -*-
+
+from __future__ import unicode_literals
+
+
+class GroupAuthFilter(object):
+    def __init__(self, request):
+        self.auth_domain = request.auth_domain
+        self.authenticated_user = request.authenticated_user
+        self.session = request.db
+        self.group_service = request.find_service(name='group')
+
+    def __call__(self, _):
+        groups = set(['__world__'])
+
+        if self.authenticated_user:
+            for group in self.authenticated_user.groups:
+                groups.add(group.pubid)
+
+        for pubid in self.group_service.world_readable_pubids():
```

Comment:
This is ok, and we can definitely merge it, but it would seem more logical for me for the group service to expose a `.groups_readable_by(user)` method and encapsulate the logic in the service rather than here. That way you could also fetch all the groups in one go:

```python
def groups_readable_by(user, query_options=None):
    readable = (Group.readable_by = ReadableBy.world)

    if user is not None:
        readable_member = sa.and_(Group.readable_by = ReadableBy.members, Group.members.has(user))
        readable = sa.or_(readable, readable_member)

    query = self.session.query(Group).filter(readable)

    if query_options is not None:
        query = query.options(*query_options)

    return query
```

...and then in the filter...

```python
groups = group_service.groups_readable_by(self.authenticated_user, 
                                          query_options=[sa.orm.load_only('pubid')])
```

etc.